File: javascript/problems/add-strings.js
/**
 * @since 2017-06-10 14:08:59
 * @author vivaxy
 *
 * @see https://leetcode.com/problems/add-strings/

 Given two non-negative integers num1 and num2 represented as string, return the sum of num1 and num2.

 Note:

 The length of both num1 and num2 is < 5100.
 Both num1 and num2 contains only digits 0-9.
 Both num1 and num2 does not contain any leading zero.
 You must not use any built-in BigInteger library or convert the inputs to integer directly.

 */

/**
 * @see https://leetcode.com/submissions/detail/105557949/
 *
 * @param {string} num1
 * @param {string} num2
 * @return {string}
 */
var addStrings = function(num1, num2) {
    var count = Math.max(num1.length, num2.length);
    var index = 0;
    var result = '';
    var remaining = 0;
    while (index++ < count) {
        var digital1 = Number(num1[num1.length - index] || '0');
        var digital2 = Number(num2[num2.length - index] || '0');
        var sum = digital1 + digital2 + remaining;
        remaining = sum > 9 ? 1 : 0;
        result = String(sum > 9 ? sum - 10 : sum) + result;
    }
    if (remaining) {
        result = String(remaining) + result;
    }
    return result;
};

var test = require('ava');
test('main', function(t) {
    t.is(addStrings('100', '100'), '200');
    t.is(addStrings('9333852702227987', '85731737104263'), '9419584439332250');
    t.is(addStrings('1', '9'), '10');
});

File: python/problems/add_strings.py
"""
https://leetcode.com/problems/add-strings/

https://leetcode.com/submissions/detail/106766561/
"""

class Solution(object):
    def addStrings(self, num1, num2):
        """
        :type num1: str
        :type num2: str
        :rtype: str
        """
        length1 = len(num1)
        length2 = len(num2)
        length = max(length1, length2)
        result = ''
        remaining = 0
        for i in range(length):
            digital1 = int(num1[length1 - 1 - i]) if length1 > i else 0
            digital2 = int(num2[length2 - 1 - i]) if length2 > i else 0
            sum = digital1 + digital2 + remaining
            remaining = 1 if sum > 9 else 0
            result = str(sum - 10 if sum > 9 else sum) + result
        if remaining != 0:
            result = str(remaining) + result
        return result

import unittest
class Test(unittest.TestCase):
    def test(self):
        solution = Solution()
        self.assertEqual(solution.addStrings('1', '2'), '3')
        self.assertEqual(solution.addStrings('100', '100'), '200')
        self.assertEqual(solution.addStrings('9333852702227987', '85731737104263'), '9419584439332250')
        self.assertEqual(solution.addStrings('1', '9'), '10')

if __name__ == '__main__':
    unittest.main()